apps/passport-server/src/database/queries/frogcrypto.ts

```diff
@@ -0,0 +1,131 @@
+import { Biome } from "@pcd/eddsa-frog-pcd";
+import {
+  FrogCryptoDbFrogData,
+  FrogCryptoFrogData
+} from "@pcd/passport-interface/src/FrogCrypto";
+import _ from "lodash";
+import { Client } from "pg";
+import { Pool } from "postgres-pool";
+import { FrogCryptoUserFeedState } from "../models";
+import { sqlQuery } from "../sqlQuery";
+
+/**
+ * Fetches state of all feeds for a particular user.
+ */
+export async function fetchUserFeedsState(
+  client: Pool,
+  semaphoreId: string
+): Promise<FrogCryptoUserFeedState[]> {
+  const result = await sqlQuery(
+    client,
+    `select * from frogcrypto_user_feeds where semaphore_id = $1`,
+    [semaphoreId]
+  );
+
+  return result.rows;
+}
+
+/**
+ * Ensure that a user has a feed specific row which we can acquire lock for.
+ *
+ * NB: This must run before transaction begins to ensure there is a lock TX can acquire and avoid deadlocks.
+ */
+export async function initializeUserFeedState(
+  client: Pool,
+  semaphoreId: string,
+  feedId: string
+): Promise<void> {
+  await sqlQuery(
+    client,
+    `insert into frogcrypto_user_feeds
+    (semaphore_id, feed_id) values ($1, $2)
+    on conflict (semaphore_id, feed_id) do nothing`,
+    [semaphoreId, feedId]
+  );
+}
+
+/**
+ * Update the last time a user has polled a feed.
+ *
+ * @returns the old last_fetched_at value or undefined if the row was not found.
+ */
+export async function updateUserFeedState(
+  client: Client,
+  semaphoreId: string,
+  feedId: string
+): Promise<Date | undefined> {
+  const result = await client.query(
+    `
+    update frogcrypto_user_feeds new
+    set last_fetched_at = $3
+    from (select * from frogcrypto_user_feeds where semaphore_id = $1 and feed_id = $2 for update nowait) old
+    where old.id = new.id
```

Comment:
not sure i fully understand this locking strategy

Reply:
clarified in the code comment